### ISSUE TITLE
Cancellable Base (e.g. Channel)

### ISSUE THREAD

**Issue by stemann** [davidanthoff/CancellationTokens.jl#14]
Would it make sense to implement some cancellable versions of Base struct's, e.g. Channel?

In this package or in another package?

**Comment by stemann** [comment]
... I'm very puzzled that is package is not used more widely/starred: Are there other options for cancellation in Julia?

**Comment by fonsp** [comment]
I would also find it super valuable to have something like `take!(channel, cancellation_token)`!

**Comment by stemann** [comment]
IIRC it’s hackable by notifying the Internal Condition of the Channel with an exception.

**Comment by stemann** [comment]
❤️  https://hackmd.io/@jakobnissen/SksGljkfkl

**Comment by stemann** [comment]
❤️ https://hackmd.io/@jpsamaroo/HylrvZDNyg

**Comment by davidanthoff** [comment]
I just left the same comment over in the spec: why start a new thing? Wouldn't it make more sense to try to improve this package here, see how much traction we can get with it and then move from there?

**Comment by stemann** [comment]
I agree that, it might be worthwhile to consider just expanding on the work in this package - going with (or at least starting with) the current implementation of the .NET perspective on cancellation.

However, I do also think that it makes sense to explore the concept in a wider perspective (looking at the perspectives of other languages etc.), and so far a HackMD document seems like a nice platform for this.

**Comment by davidanthoff** [comment]
In my mind the best way forward with something like this is to actually exercise some code in the wild and see how it goes, iterate, and then at some point if there has been some critical mass of usage one can think about moving it to base.

I've started to use this package (finally!) for real now in the VS Code extension ecosystem. I just published a new version of JSONRPC.jl that handles cancellation tokens across JSONRPC calls, and so far it all is looking pretty good.

But really I think the most useful thing would be to actually add some of the things that are missing in this package, like this issue here, and also new methods for many of the base IO functions.
